Implemented property compatibilityChangeFilterFile for filter required compatibility changes by yml file.

Pochatkin opened the pull request:
1. Java 7 -> 8.
2. TODO: support version validate
3. TODO: Try to generify Change*Entity and Remove*Entity 
4. TODO: More test and docs?

Commit:
Implemented property compatibilityChangeFilterFile for filter required compatibility changes by yml file.

Cyrill commented:
Please add `JapicmpTask = me.champeau.gradle.japicmp.JapicmpTask.class` to `JapicmpPlugin.apply`

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JapicmpTask.java:
```diff
@@ -168,18 +171,44 @@ private void addClasspathFor(Class<?> clazz, Set<File> classpath) {
+  }
+
+
+  public static void main(String[] args) {
+    String s = tryExtractVersion(new File("test1-1.1.jar"));
+    assert s.equals("1.1");
+    String s1 = tryExtractVersion(new File("test1-1.2-SNAPSHOT.jar"));
+    assert s1.equals("1.2-SNAPSHOT");
+  }
```
Please use tests if you need to check class behavior

Pochatkin replied:
This is just trash which I forgot remove.

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JapicmpTask.java:
```diff
@@ -463,4 +492,14 @@ public void setRichReport(RichReport richReport) {
   }
 
+  @Optional
+  @Nested
```
It is not a Nested property. Please annotate as `@InputFile` instead
```suggestion
  @InputFile
  @PathSensitive(PathSensitivity.RELATIVE)
```

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JapiCmpWorkerConfiguration.java:
```diff
@@ -51,60 +51,63 @@
+                                      List<Archive> newArchives,
+                                      boolean onlyModified,
+                                      boolean onlyBinaryIncompatibleModified,
+                                      boolean failOnSourceIncompatibility,
```
Inconsistent formatting

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JApiCmpWorkerAction.java:
```diff
@@ -197,9 +201,19 @@ private void generateOutput(JarArchiveComparator jarArchiveComparator) {
+      CompatibilityChangesFilter filter = new CompatibilityChangesFilter(compatibilityChangesFilterFile);
+      Set<String> collect = oldArchives.stream().map(Archive::getFileName)
+          .collect(Collectors.toSet());
+      Set<String> collect1 = newArchives.stream().map(Archive::getFileName)
```
I am sure we are able to find a better name for this variable

Commit:
Code cleanup and fix new task arg settings

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JApiCmpWorkerAction.java:
```diff
@@ -197,9 +201,19 @@ private void generateOutput(JarArchiveComparator jarArchiveComparator) {
+      Set<String> collect = oldArchives.stream().map(Archive::getFileName)
+          .collect(Collectors.toSet());
+      Set<String> collect1 = newArchives.stream().map(Archive::getFileName)
+          .collect(Collectors.toCollection(() -> collect));
```
Did I get it right that the code here simply does this?
```
 Stream.concat(
          oldArchives.stream().map(Archive::getFileName),
          newArchives.stream().map(Archive::getFileName))
    .collect(toSet());
```

Pochatkin replied:
Yeap, done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/Archive.java:
```diff
@@ -0,0 +1,39 @@
+    String[] nameAndExtension = Parser.splitByLastDotChar(name);
+    String s = nameAndExtension[0];
+    int i = s.lastIndexOf(version);
+    return s.substring(0, i - 1);
```
Imagine we have a `foo-common-2.22-jdk8.jar`, at the end of this method we take all chars of file name from the beginning up to the char before the version part. 
Why do we call `Parser.splitByLastDotChar(name)` at all?

And what if the code that has instantiated a new instance of this class failed to get version? `JapicmpTask.tryExtractVersion` returns `1.0`  in case of exception and `i` would become `-1` here.

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JapicmpTask.java:
```diff
@@ -168,18 +171,44 @@ private void addClasspathFor(Class<?> clazz, Set<File> classpath) {
+    } catch (Exception ignored) { }
+    return "1.0";
+  }
+
```
I would prefer to see a test for this

Pochatkin replied:
Of course. Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,155 @@
+        }
+      }
+    }
+    System.out.println(projects);
```
Let's use Gradle loggers instead. See `DefaultTask` parents for details how to add it

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,155 @@
+        for (Map<String, String> entity : value) {
+          project.add(extractEntity(entity));
+        }
+      }
```
Please extract project creation to a separate method

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,155 @@
+      throw new NullPointerException();
+    } else if (entity == null) {
+      throw new NullPointerException();
+    }
```
Please add a message to the exceptions

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/entity/Entity.java:
```diff
@@ -0,0 +1,12 @@
+    this.version = version;
+  }
+
+  public abstract String getClassName();
```
Looks like all subclasses have exactly the same implementation

Pochatkin replied:
Yeap, it always calls Element.getCanonicalName, but we do not have access to element field here.

Cyrill replied:
Every subsclass is initialized with at least one final `Element`. If we add `Element` to `Entity` constructor, we won't need to duplicate the implementation of `getClassName()`

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/entity/Entity.java:
```diff
@@ -0,0 +1,12 @@
+
+//TODO: Implement version supporting
+public abstract class Entity {
+  private final String version;
```
private final, no accessor?

Pochatkin replied:
I not supported version checking yet, I will push new changes about it soon.

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/entity/ClassDescriptor.java:
```diff
@@ -0,0 +1,135 @@
+
+  public void add(Entity entity) {
+    if (entity instanceof RemoveMethodEntity) {
+      RemoveMethodEntity castedEntity = (RemoveMethodEntity) entity;
+      removedMethods.put((castedEntity).getName(), castedEntity);
+    } else if (entity instanceof ChangeMethodEntity) {
+      ChangeMethodEntity castedEntity = (ChangeMethodEntity) entity;
+      changedMethods.put(castedEntity.getName(), castedEntity);
+    } else if (entity instanceof RemoveFieldEntity) {
+      RemoveFieldEntity castedEntity = (RemoveFieldEntity) entity;
+      removedField.put((castedEntity).getName(), castedEntity);
+    } else if (entity instanceof ChangeFieldEntity) {
+      ChangeFieldEntity castedEntity = (ChangeFieldEntity) entity;
+      changedFields.put((castedEntity).getFirstFieldName(), castedEntity);
+      changedFields.put((castedEntity).getSecondFieldName(), castedEntity);
+    } else if (entity instanceof RemoveConstructorEntity) {
+      RemoveConstructorEntity castedEntity = (RemoveConstructorEntity) entity;
+      removedConstructors.put((castedEntity).getName(), castedEntity);
+    } else if (entity instanceof ChangeConstructorEntity) {
+      ChangeConstructorEntity castedEntity = (ChangeConstructorEntity) entity;
+      changedConstructors.put((castedEntity).getName(), castedEntity);
+    } else if (entity instanceof RemoveClassEntity) {
+      RemoveClassEntity castedEntity = (RemoveClassEntity) entity;
+      removeClasses.put(castedEntity.getClassName(), castedEntity);
+    }
```
Any way to avoid `instanceof`?

Pochatkin replied:
I would like to leave this as a technical debt for now. Didn't find easy way to avoid it.

Cyrill replied:
Well, how about a Visitor?

Commit:
Code format, exception messages, better stream API using

Commit:
Write tests for version parser

Cyrill commented on src/test/java/me/champeau/gradle/japicmp/ignore/ParserTest.java:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  @Test
+  public void testVersionExtracter() {
```
This test fails if a jar has a classifier 
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.bundling.Jar.html#org.gradle.api.tasks.bundling.Jar:archiveFileName

Here is a live example of it `https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-hdfs/3.3.0/hadoop-hdfs-3.3.0-tests.jar`

Pochatkin replied:
Done

Commit:
1. Implemented support for version check.
2. Unify version class.
3. Implemented tests for check version property
4. Fix problems with jar name parse

Commit:
Minor changes

Commit:
Unify RemoveEntity and ChangeEntity. Removed codeduplicates

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JApiCmpWorkerAction.java:
```diff
@@ -197,9 +201,14 @@ private void generateOutput(JarArchiveComparator jarArchiveComparator) {
     Options options = Options.newDefault();
     options.setOldClassPath(japicmp.util.Optional.of(toClasspath(oldClasspath)));
     options.setNewClassPath(japicmp.util.Optional.of(toClasspath(newClasspath)));
-    final List<JApiCmpArchive> baseline = toJapiCmpArchives(oldArchives);
```
toJapiCmpArchives is not used anymore

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/archive/Version.java:
```diff
@@ -0,0 +1,145 @@
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      if (era != null) {
+        sb.append(era);
+      }
+      if (major != null) {
+        sb.append(".").append(major);
+      }
+      if (minor != null) {
+        sb.append(".").append(minor);
+      }
+      if (patch != null) {
+        sb.append(".").append(patch);
+      }
+      return sb.toString();
```
What is expected when era=null, major=1, minor=1, patch=null?

Pochatkin replied:
Done

Commit:
Unify version class. Another minor changes

Commit:
Added JapicmpTask to extension extra properties of working project

Pochatkin commented:
> Please add `JapicmpTask = me.champeau.gradle.japicmp.JapicmpTask.class` to `JapicmpPlugin.apply`

Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/archive/Version.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      for (int element : elements) {
+        sb.append(element).append(".");
+      }
+      if (elements.length > 0) {
+        sb.deleteCharAt(sb.length() - 1);
+      }
+      return sb.toString();
+    }
```
Did you mean this?
```suggestion
    public String toString() {
      return Arrays.stream(elements)
          .mapToObj(String::valueOf)
          .collect(joining("."));
    }
```

Cyrill replied:
Or you can take `StringJoiner` for this if stream is not what you prefer

Pochatkin replied:
Yeap, thanks!

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/archive/Version.java:
```diff
@@ -0,0 +1,123 @@
+      return sb.toString();
+    }
+
+    public static SemanticVersion semanticVersion(String rawVersion) throws VersionParseException {
+      if (rawVersion.endsWith(".")) {
+        throw new VersionParseException("Incorrect version format with last dot char: " + rawVersion);
+      }
+      String[] split = rawVersion.split("\\.");
+
+      List<Integer> result = new ArrayList<>();
+      for (String s : split) {
+        try {
+          int i = Integer.parseInt(s);
+          result.add(i);
+        } catch (Exception e) {
+          break;
+        }
+      }
+
+      if (result.isEmpty()) {
+        throw new VersionParseException("The input: " + rawVersion + " can't be parse as version");
+      }
+
+      return new SemanticVersion(result.stream().mapToInt(i->i).toArray());
+    }
```
How about this?
```suggestion
private static OptionalInt parseSection(String section) {
      try {
        return OptionalInt.of(Integer.parseInt(section));
      } catch (Exception e) {
        return OptionalInt.empty();
      }
    }
    
 public static SemanticVersion semanticVersion(String rawVersion) throws VersionParseException {
      if (rawVersion.endsWith(".")) {
        throw new VersionParseException("Incorrect version format with last dot char: " + rawVersion);
      }
      int[] versionSections = Arrays.stream(rawVersion.split("\\."))
          .map(SemanticVersion::parseSection)
          .filter(OptionalInt::isPresent)
          .mapToInt(OptionalInt::getAsInt)
          .toArray();

      if (versionSections.length == 0) {
        throw new VersionParseException("The input: " + rawVersion + " can't be parse as version");
      }
      return new SemanticVersion(versionSections);
    }
```

Pochatkin replied:
Yeah, cool!

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/archive/Archive.java:
```diff
@@ -0,0 +1,43 @@
+  }
+  
+  
+  public String getFileName() {
+    String name = file.getName();
+    try {
+      int i = name.lastIndexOf(version.toString());
+      return name.substring(0, i - 1);
+    } catch (Exception ignored) {}
+    return name;
+  }
```
We can avoid parsing file name if we create this class with it.
`resolvedDependency.getAllModuleArtifacts().iterator().next().getName()` may be what is expected

Pochatkin replied:
But we also have the possible instantiation this by just FileCollection. In this case we will have full file name. 

```
private static List<Archive> toArchives(FileCollection fc) {
    Set<File> files = fc.getFiles();
    List<Archive> archives = new ArrayList<>(files.size());
    for (File file : files) {
      archives.add(new Archive(file, Parser.tryExtractVersion(file)));
    }
    return archives;
  }
```

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/Project.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Project {
```
Please give a different name to this class. It's easy to misread it as org.gradle.api.Project

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,161 @@
+
+    return new RemoveEntity<>(version, element);
+
+//    if (element instanceof FieldElement) {
+//      return new RemoveFieldEntity(version, (FieldElement) element);
+//    } else if (element instanceof MethodElement) {
+//      return new RemoveMethodEntity(version, (MethodElement) element);
+//    } else if (element instanceof ConstructorElement) {
+//      return new RemoveConstructorEntity(version, (ConstructorElement) element);
+//    } else if (element instanceof ClassElement) {
+//      return new RemoveClassEntity(version, (ClassElement) element);
+//    }
+
+//    return null;
```
Do we need this code?

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JApiCmpWorkerAction.java:
```diff
@@ -197,9 +193,14 @@ private void generateOutput(JarArchiveComparator jarArchiveComparator) {
+    Diff diff = new Diff(jarArchiveComparator, oldArchives, newArchives);
+    List<JApiClass> jApiClasses = diff.classes();
+
+    if (compatibilityChangesFilterFile != null) {
+      CompatibilityChangesFilter filter = new CompatibilityChangesFilter(compatibilityChangesFilterFile);
+      filter.filterChanges(diff);
+    }
```
Do I get it right that this piece of code actually modifies instances of `JApiClass`, so `List<JApiClass> jApiClasses = diff.classes();` is affected?

Pochatkin replied:
Yeah, this code modify the instances of JApiClass but the LIST is not modified.

Commit:
Fixed review comment. Code cleanup

Commit:
Minor improvemetns

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/archive/Version.java:
```diff
@@ -0,0 +1,118 @@
+    return true;
+  }
+
+  private static class SemanticVersion implements Comparable<SemanticVersion>, Serializable {
```
Do we really need this inner class? Can't we merge it with Version?

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/JApiCmpWorkerAction.java:
```diff
@@ -197,9 +191,11 @@ private void generateOutput(JarArchiveComparator jarArchiveComparator) {
-    final List<JApiCmpArchive> current = toJapiCmpArchives(newArchives);
-    List<JApiClass> jApiClasses = jarArchiveComparator.compare(baseline, current);
+    Diff diff = new Diff(jarArchiveComparator, oldArchives, newArchives);
+    List<JApiClass> jApiClasses = compatibilityChangesFilterFile != null
+        ? new CompatibilityChangesFilter(compatibilityChangesFilterFile).filterChanges(diff)
+        : diff.classes();
```
Every time I see this block of code it makes me feel that something is not quite right. And I think I have finally realized what it is.
Here we have a conditional statement, but in fact both branches call `diff.classes()`, which is not clear at all when you look at it.
Something like this better describes what we really expect from this code
```
List<JApiClass> jApiClasses = diff.classes();
if(compatibilityChangesFilterFile != null){
  jApiClasses =  new CompatibilityChangesFilter(compatibilityChangesFilterFile).filterChanges(jApiClasses)
}
```

Pochatkin replied:
Done

Cyrill replied:
Well, not really done. Maybe I was not making myself clear, but we now call diff.classes() twice, in JApiCmpWorkerAction and in CompatibilityChangesFilter. So the latter is not really filtering anything, but rather building the list in a different way. So the responsibility to build the list of classes is shared between JApiCmpWorkerAction and in CompatibilityChangesFilter.
The root cause of this issue is that `Diff` is actually a mutable one. 
```
List<JApiClass> classesA = diff.classes();
List<JApiClass> classesB =  new CompatibilityChangesFilter(compatibilityChangesFilterFile).filterChanges(diff);
List<JApiClass> classesC =  diff.classes();
```
makes `classesC` equal to `classesB` and not `classesA`

Cyrill replied:
We can leave it as tech debt for now

Cyrill commented on gradle/compile.gradle:
```diff
@@ -23,10 +23,10 @@ repositories {
-        exclude group:'io.airlift'
-        exclude group:'javax.xml.bind'
-        exclude group:'com.sun.xml.bind'
+//        exclude group:'com.google.guava'
+//        exclude group:'io.airlift'
+//        exclude group:'javax.xml.bind'
+//        exclude group:'com.sun.xml.bind'
```
Why is it commented out?

Pochatkin replied:
I see no reason at all to make exclusions for this.

Cyrill replied:
then please remove those lines at all

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,143 @@
+    declaredModules = fillProjects(objects);
+  }
+
+  private List<DeclaredModule> fillProjects(HashMap<String, List<Map<String, String>>> objects) {
```
fillModules?

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,143 @@
+    return result;
+  }
+
+  private DeclaredModule parseAsModule(Map.Entry<String, List<Map<String, String>>> projectEntry) {
```
Meaningful parameters look more conceivable
```suggestion
private DeclaredModule parseAsModule(String name, List<Map<String, String>> values) {
```

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/entity/Entity.java:
```diff
@@ -0,0 +1,19 @@
+
+import me.champeau.gradle.japicmp.archive.Version;
+
+public abstract class Entity<T> {
```
I would say this is rather `Action` or `Modification` than `Entity`

Pochatkin replied:
This naming is come from yml config file structure. If we want rename it I think we also should rename prop names in yml file.

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/entity/EntityManager.java:
```diff
@@ -0,0 +1,49 @@
+    }
+  }
+
+  public boolean validate(Provider provider, VersionsRange versionsRange) {
```
This method in fact does not validate anything, it checks if Element matches Entity.
Maybe better rename this and all affected methods  to `matches` or something like that?

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/processor/ClassProcessor.java:
```diff
@@ -0,0 +1,72 @@
+        switch (compatibilityChange) {
+          case CLASS_REMOVED:
+            classMutator.removeCompatibilityChange(clazz, JApiCompatibilityChange.CLASS_REMOVED);
+        }
```
'switch' statement has too few case labels and should probably be replaced with an 'if' statement

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/CompatibilityChangesFilter.java:
```diff
@@ -0,0 +1,143 @@
+
+
+  public List<JApiClass> filterChanges(Diff diff) {
+    List<Entity<?>> collect = declaredModules.stream()
+        .filter(diff::containsProject)
+        .map(DeclaredModule::getEntities)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
```
Please disregard this comment if I got this wrong, but as far as I see what is happening here, we
 iterate all `declaredModules`, which are jar names in fact. Then we collect all modifications(entities) from all projects into a single list.
In an imaginary case we could have two classes in different jars with the same canonical name, in that case it won't be possible to tell which modification(entity) belongs to which jar. 
I admit this is unlikely to happen in prod, since current requirements expect one jar (declaredModule) per yml file, nevertheless it is still possible to mess up

Pochatkin replied:
Yes, you are right, we will not know which entity corresponds to each jar. BUT this is not need, because in the state of matching, we try to match each with each, and we will definitely find the required pair, if it exists.

Cyrill replied:
In that case the plugin might mistakenly report no issues when there are some. Imagine, we have two modules, xxx-adk-server.jar and xxx-adk-client.jar both having a class named `com.wandisco.serialization.Format`. A developer was going to change only the client implementation, but accidentally IDEA refactoring updated the server one as well, and what's more it was overlooked on review. So the change in the server jar is unwanted and expected to be detected by this plugin. Will it raise a flag?

Pochatkin replied:
Yeah, you are right. But I think this is very rare case because we ussialy provide only one jar file per task. Do you really think that we should fix it right in this pull request?

Cyrill replied:
Well this case at least deserves a comment. We can leave it as is for now

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/processor/ClassProcessor.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  private void doProcessClass(JApiClass clazz, VersionsRange versions) {
+    provider.setRemoveElement(clazz);
```
I could agree that this approach is legit if we were limited on memory and, moreover, the other parts of this plugin were memory-optimized as well, but it is not our case.

Please use wrapper classes with a particular strategy instead of injecting objects into strategies

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/processor/ClassProcessor.java:
```diff
@@ -0,0 +1,72 @@
+    constructorProcessor = new ConstructorProcessor(classMutator, manager);
+  }
+
+  public void process(List<JApiClass> classes, VersionsRange versions) {
+    for (JApiClass clazz : classes) {
+      doProcessClass(clazz, versions);
+      methodProcessor.process(clazz.getMethods(), versions);
+      fieldProcessor.process(clazz.getFields(), versions);
+      constructorProcessor.process(clazz.getConstructors(), versions);
+      classMutator.tryClearClass(clazz);
+    }
+  }
```
It looks like it's not this class responsibility to own this method. `ClassProcessor` should just operate with class modifications, pretty much the same way methodsProcessor or fieldProcessor do.

Pochatkin replied:
Done

Cyrill commented on src/main/java/me/champeau/gradle/japicmp/ignore/processor/Changer.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Map;
+import java.util.function.Function;
+
+public class Changer<T, R> {
```
Javadocs would drastically reduce the time spent on reading the code

Pochatkin replied:
Done

Commit:
Review changes

Commit:
minor

Cyrill reviewed (approved):
LGTM
